contracts/margin/external/exchangewrappers/ZeroExV2MultiOrderExchangeWrapper.sol

```diff
@@ -205,30 +251,78 @@ contract ZeroExV2MultiOrderExchangeWrapper is
                 );
             }
 
+            // ignore orders that the maker will not be able to fill
+            if (!makerHasEnoughTokens(
+                makerToken,
+                balances,
+                order.makerAddress,
+                available.makerAmount)
+            ) {
+                continue;
+            }
+
             // update the running tallies
             total.takerAmount = total.takerAmount.add(available.takerAmount);
             total.makerAmount = total.makerAmount.sub(available.makerAmount);
         }
+    }
 
-        // require that all amount was bought
-        require(
-            total.makerAmount == 0,
-            "ZeroExV2MultiOrderExchangeWrapper#getExchangeCost: Cannot buy enough maker token"
-        );
+    /**
+     * Checks and modifies balances to keep track of the expected balance of the maker after filling
+     * each order. Returns true if the maker has enough makerToken left to transfer amount.
+     */
+    function makerHasEnoughTokens(
+        address makerToken,
+        TokenBalance[] memory balances,
+        address makerAddress,
+        uint256 amount
+    )
+        private
+        view
+        returns (bool)
+    {
+        // find the maker's balance in the cache or the first non-populated balance in the cache
+        TokenBalance memory current;
+        uint256 i;
+        for (i = 0; i < balances.length; i++) {
+            current = balances[i];
+            if (
+                current.owner == address(0)
+                || current.owner == makerAddress
+            ) {
+                break;
+            }
+        }
 
-        // validate that max price will not be violated
-        validateTradePrice(
-            priceRatio,
-            total.takerAmount,
-            desiredMakerToken
-        );
+        // if the maker is already in the cache
+        if (current.owner == makerAddress) {
+            if (current.balance >= amount) {
```

Comment:
I'm confused by this. I don't see `current.balance` even being set anywhere

Reply:
Think of `current` as a pointer rather than a copy-by-value

Reply:
but I'm saying it doesn't look like `current.balance` is even initialized. So won't it always be 0?

The only place I see that it's modified is `current.balance = current.balance.sub(amount);`